File: 06_assoc_prop.rb
# frozen_string_literal: true

scream  = ->(str) { str.upcase }
exclaim = ->(str) { "#{str}!" }
repeat  = ->(str) { "#{str} #{str}" }

compose = ->(*fns) {
  ->(x) {
    fns.reverse_each.reduce(x) { |acc, fn| fn.(acc) }
  }
}

with_exuberance = compose.(repeat, exclaim, scream)
repeat_excitedly = compose.(repeat, exclaim)

str = 'String'

puts with_exuberance.(str)
puts compose.(repeat_excitedly, scream).(str)

File: 06_assoc_prop.js
const { compose, scream, exclaim, repeat } = require('./shared')

const withExuberance = compose(repeat, exclaim, scream)

const repeatExcitedly = compose(repeat, exclaim)

const str = 'String'

console.log(withExuberance(str))
console.log(compose(repeatExcitedly, scream)(str))
